### ISSUE TITLE
[0.9.1.9] Server Discovery is Inconsistent in the Server Browser

### ISSUE THREAD

**Issue by MonzUn** [StrangeLoopGames/EcoIssues#20071]
There have been some posts the recent month(s) on the official Discord about users not being able to see servers in the Server Browser and as to my knowledge, they haven't received any tangible help with this.

My own server has been affected by this and Nemo, me and Mishka have been investigating the issue.
It usually takes about 30 refreshes on the server browser for Nemo or I before can see the server in the list.
This means that the server is getting basically no new players from the server browser. For us who have a relatively stable community, it's not a critical issue, but if this would have happened when we started a year ago, we would likely have just quit, so it more than likely has a negative effect on freshly started servers.

We have tried a lot of fiddling with the server settings and the network setup, but right now it's looking like it is an issue with the server responsible for keeping the server lists. The test that made suspect it's master server being wrong is that we found out that one of Mishka's servers (snepserv) always shows up so we tested to put a server with my config on the same hardware and the result was that it still had the same behavior of only showing very rarely.

You should be able to test this with Sea Otter and White Tiger, as Otter has this issue while Tiger does not.

More things we have tried:
* Fresh server install with no mods.
* Not being behind NAT.
* Many combinations of server settings.

Test Server Config (copied from my server. ID and Passport freshly generated and IP changed to match Mishka's network)
```
{
  "PublicServer": true,
  "Playtime": "",
  "DiscordAddress": "https://discord.gg/9CANEWy",
  "Password": "testing",
  "Description": "EcoWorld -  <color=#FFFF00>Anniversary  -  Extra Collaboration</color>  -  <color=#1E90FF>Beginners Welcome</color>",
  "DetailedDescription": "Welcome!\nEcoworld is a casual server where beginners are welcomed with open arms! We have been running for a year now and the current cycle is our anniversary! \n\nFeatures:\n- Stable dedicated 24/7 server\n- Discord server with game integration.\n- Open and welcoming community.\n- Dedicated mods with experience of Eco 9.0.\n - Regular backups.-Mods for extra collaboration and utility\n    - DiscordLink\n    - Chronicler\n    - Ucat's price calculator\n    - No More Books\n    - Furniture M3\n    - Elixir mods\n  - Nids Toolkit",
  "ServerCategory": "Beginner",
  "IPAddress": "Any",
  "RemoteAddress": "xx.xxx.xx.xxx",
  "WebServerUrl": "http://xx.xxx.xx.xxx:5001/",
  "GameServerPort": 5000,
  "WebServerPort": 5001,
  "Rate": 25,
  "MaxConnections": -1,
  "ID": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
  "Passport": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
  "UPnPEnabled": true
}
```

These two images are taken with only a refresh in between and show the top 4 servers sorted for population.
![Sea Otter Found](https://user-images.githubusercontent.com/11805935/104970161-d8d52a00-59ea-11eb-88b4-87e60216ee68.png)
![Sea Otter Lost](https://user-images.githubusercontent.com/11805935/104970163-da9eed80-59ea-11eb-9b63-7bb45fecebe3.png)

Feel free to reach out to me here or on Discord if you have questions or want any help with testing this issue.

**Comment by MishkaFur** [comment]
As tested on my hardware that runs SnepServ (Server 2016 VM) which always appears

Copied my servers entire folder to a new folder, put in Monzun's network.eco and set it to use port 5000 and 5001 respectively. server only once showed for me in the listing.

Port forwarding and firewall settings are identical for my server (port 4000/4001 and the test one for Monzun

Official servers are VERY intermittent for me to show in the browse all list which are known to have correctly working configs (Sea Otter almost never appears but Coast Redwood appears easily), I have tested this on 3 different locations in the UK on 3 different ISP's

![image](https://user-images.githubusercontent.com/61889138/104970366-1cc43100-59e3-11eb-840e-757ec4683a23.png)

As i expect it'll be Mirasrael looking into this, prod me on discord if I can provide any info/access onto my hardware for further testing.

**Comment by evanah64** [comment]
I'm glad someone is testing this. We have had this very issue a few times and it's essentially a server/season killer. Ours happens to be a modded server and it launched within an hour of TED TOWN shown in the above screen shot. They also run a modded server with very similar settings to ours. Theirs showed up on New Beginnings right away as well as the top of the browser list and ours did not show up anywhere until day 2 into day 3 even with dozens of refreshes on the browser list. By that time it's to late for people looking for new servers to join. We do have an established server so it was even 'seeded' with admins and those from our community that were able to join using the IP. This did not seem to make any difference. For contrast we ended up with about 30 total log ins compared to the 300+ of the TED server. We usually see 4-5 times that number.

**Comment by NemoLArs** [comment]
Like Monz said in his post, we have tried everything we could think about. The thing that bugs me and confuse me is that our servers get listed 1/20 times ish. What changes between these refreshes, either everytime or never would be more reasonably.

I've talked to server owners who has "always-show" server and got their configs / setup etc, tried it on my server and got a "no-show" server. 

Tried Windows Server, Windows 10 and linux (no good result on linux... ) 

Tried a huge port range. 
Same ports as always-show servers. 

I've sent packages and pinged always-shown servers and no-show server to see if i have any issues on my end, results were the same on any server really. 

Conclusion is after, without to exxagerate 30 hours of troubleshooting:
The issue seems to be on Eco's end and not the users. 

If you need more information or any help testing stuff just PM me on discord, Nemo - Mr Magic is my name.

**Comment by StallEF** [comment]
this problem was identified back in May 2020 on version 8.3.3. 

1) the list of servers running for less than 10 days works correctly, but a different algorithm is used for servers longer than 10 days
2) the display of servers is dramatically affected by the server category specified in the setting
3) if you apply a filter and you should have only, say, 3 servers, then with each update there will be 1 or 2 or 3 absolutely randomly
4) we were promised this would be fixed in 9.0

My opinion on this is that the server sends the client a complete list of servers, and the client is already sorting by filter. But since the server probably has a function to interrupt waiting for a response from the server about its characteristics, the response sent to the client is often incomplete.

**Comment by MishkaFur** [comment]
> 
> 
> this problem was identified back in May 2020 on version 8.3.3.
> 
>     1. the list of servers running for less than 10 days works correctly, but a different algorithm is used for servers longer than 10 days

From I can see its the reverse, longer running servers show but newer ones do not ie EcoWorld and Sea Otter though some exceptions apply, I cannot see the australia Oceanic server in the list at all (server done by IronyTroll does not appear and its on day 47), WT does appear and its on day 39

Judging by the servers that do not appear though or at least very occasionally do, its having a big impact on playerbase. looking through the list of servers I can almost assume what servers do often show up vs what ones do not due to the player count which will dishearten new server owners because not getting many players despite the settings seemingly being perfect.

**Comment by MonzUn** [comment]
I agree with Mishka's observation here.

```
the display of servers is dramatically affected by the server category specified in the setting
```
In what way did the server category affect the results? From my own testing, it doesn't seem to affect the result much at all.

**Comment by evanah64** [comment]
I can also say it appears to be the reverse. This is common when the server is new (most critical time) and seems to be less of an issue after 2 to 3 days.

**Comment by MishkaFur** [comment]
Scrolling down through server list, clicking "Load more" till no more load.

EcoWorld is not appearing at all and in theory my game has a complete list of every server in eco at the moment (due to no sorting by name I sorted by online count and tried to find them), doing a search just then reloads the list doesnt base it off what my game/PC has already found

doing browse all with filter of 1+ players online I cannot see Sea Otter nor coast redwood and as mentioned by others its random if things appear 

![image](https://user-images.githubusercontent.com/61889138/105054177-97d02a80-5a69-11eb-886c-bc971ca2f109.png)


Each refresh was done seconds after taking the pic

First refresh, left until grey cog is not spinning
![image](https://user-images.githubusercontent.com/61889138/105053791-3f992880-5a69-11eb-80f7-f5269d76ddb8.png)

Second refresh, where did that french server come from? and where did CanadianEh go?
![image](https://user-images.githubusercontent.com/61889138/105053835-49bb2700-5a69-11eb-867a-e2e0ec6ed2df.png)

Third refresh, where did said french server go?
![image](https://user-images.githubusercontent.com/61889138/105053978-6bb4a980-5a69-11eb-9366-0b5e88ab6cbd.png)

No official servers showing despite the fact they should have (Also note i am missing Panda)
![image](https://user-images.githubusercontent.com/61889138/105054294-bd5d3400-5a69-11eb-8579-7000f2576940.png)

**Comment by StallEF** [comment]
https://github.com/StrangeLoopGames/EcoIssues/issues/18742

**Comment by D3nnis3n** [comment]
By Phlo:.
For some time now, maybe 2-3 months, my server isn't appearing in server search results. Was initially thinking its something on my end, but it might not be.
On my end everything seems normal,players can connect easily with DNS, IP or by joining from favorites, network is working fine. The server is old, so its ID was not changed since ID became a thing; maybe that has something to do with it?!
Server DNS: eco.greenleafserver.com:3011
Server is vanilla, so no mods could of messed with it.

**Comment by mirasrael** [comment]
The issue with server not appearing in list already fixed in 9.2.

**Comment by D3nnis3n** [comment]
Fixed by https://github.com/StrangeLoopGames/Eco/pull/7124.

The algorithm preferring servers with less players starting after a threshold is intended to even out chances of less popular servers a bit.
